pandas/_libs/internals.pyx

```diff
@@ -890,17 +890,29 @@ cdef class BlockValuesRefs:
     """
     cdef:
         public list referenced_blocks
+        public int dead_counter
+        object __weakref__
+        object _weakref_cb
 
     def __cinit__(self, blk: Block | None = None) -> None:
+        def _weakref_cb(item: weakref.ref, selfref: weakref.ref = weakref.ref(self)) -> None:
+            self = selfref()
+            if self is not None:
+                self.dead_counter += 1
```

Comment:
Sorry, can you explain a bit more? Is pandas preparing to be thread-safe and use the no-GIL feature of the coming python version?

Reply:
We release the GIL in quite a few places in the Cython code today, mostly for downstream libraries like Dask to take advantage of. Though I guess this in its current state wouldn't work without the GIL since it is already using `weakref` on main